_posts/2020-09-24-Using-Zeek-to-find-persistent-threats-by-monitoring-DNS.md

```diff
@@ -0,0 +1,222 @@
+
+* Open the script in a text editor. On SensorFleet environment, you can use the Rule Manager UI for editing. On Generic environment, locate anomalous-dns package in the Zeek installation, and open domain-whitelist.zeek in a text editor.
+* Edit the last few lines, they look like this:
+  ```
+  redef domain_whitelist = /.../;
+  redef local_dns_servers = set(...);
+  redef totally_ignored_ips = set(...);
+  ```
+
+* Add any domains or DNS servers to appropriate whitelists. I ended up adding some DNS and mail servers to _totally_ignored_ips_ field to silence them, and a few common domains in _domain_whitelist_. There are a few syntax samples in the whitelist. For more details, see [Zeek scripting manual for set, addr and pattern](https://docs.zeek.org/en/current/examples/scripting/#data-structures).
+* Save script changes and restart Zeek:
+  * On SensorFleet, just re-save the script and Commit script changes.
+  * On General environment, save the script and run `/etc/init.d/zeek restart` (depending on the platform).
+
+## Step 6: Does it really work?
+
+Once you have the script running, and it’s not creating too much false positives anymore, you may want to see if it really can detect persistent DNS tunnels. Let’s test it.
+
+* For testing, you will need a Linux or other unix shell, or a Windows PowerShell capable shell in your monitored network that has the host command (or nslookup on windows). The host must not be whitelisted, of course.
+
+* Command to run in unix shell: (bash-style syntax)
+  ```
+  for n in {1..100}; do host dnstest$n.sensorfleet.com 8.8.4.4; done
+  ```
+
+* Similar test command for Windows powershell:
+  ```
+  for (($n = 0); $n -lt 100; $n++) { nslookup "dnstest$i.sensorfleet.com" 8.8.4.4 }
+  ```
+
+* The test should result in an event like this:
+
+  <img src="/img/posts/zeek_1_blog/12_test_event.png" title="Triggered test event" width=500>
+
+  Here, the id.orig_h (or src field) is the source of our potential threat.
+
+## Final words
+
+You will now get a notification of anomalous DNS behaviors, such as long running malware C&C connections over DNS. This can be fairly useful if your network is restricted (e.g. HTTP proxy), but allows direct DNS access.
+
+Also, some malwares might use DNS by default, even if HTTP was allowed, because DNS might be harder to detect since they do not make direct connections to the outside world, but instead leverage your DNS resolver.
+
+I found out that configuring the whitelist for nearly zero false positives can take a lot of work, though, depending on your network. That’s why I ended up patching the [jbaggs/anomalous-dns](https://github.com/jbaggs/anomalous-dns) for a more aggressive whitelist.
+
+We now use the above detection method in monitoring of our own infrastructure.
+
+Hopefully this will help someone to realize how neat things you can do just by using open source tools such as Zeek.
+
+It gets even easier when using commercial products to round the rough edges a bit.
```

Comment:
"  ... source tools such as Zeek. And it gets even easier..."

Combining one-sentence paragraphs.